src/main/c/NativeModule.cpp

```diff
@@ -0,0 +1,713 @@
+    fprintf(f, "CXXFLAGS  := \\\n");
+    fprintf(f, "  %s -fPIC -march=native -fno-strict-aliasing -Wall \\\n", cxxstd);
+    auto java_include = strip_suffix(config.java_home_, "/jre") + "/include";
+    fprintf(f, "  -I%s \\\n", java_include.c_str());
+    fprintf(f, "  -I%s/%s \\\n", java_include.c_str(), config.java_md_.c_str());
+    fprintf(f, "  -I%s \\\n", include_.c_str());
+    bool have_oflag = (strstr(options_.c_str(), "-O") != nullptr);
+    fprintf(f, "  %s%s \\\n", have_oflag ? "" : "-O3", options_.c_str());
+    fprintf(f, "  -DHAIL_MODULE=$(MODULE)\n");
+    fprintf(f, "LIBFLAGS := -fvisibility=default %s\n", 
+      config.is_darwin_ ? "-dynamiclib -Wl,-undefined,dynamic_lookup"
+                         : "-rdynamic -shared");
+    fprintf(f, "\n");
+    // top target is the .so
+    fprintf(f, "$(MODULE_SO): $(MODULE).o\n");
+    fprintf(f, "\t-[ -f $(MODULE).lockX ] || /usr/bin/touch $(MODULE).lockX ; \\\n");
+    fprintf(f, "\t  while ! /bin/ln $(MODULE).lockX $(MODULE).lock 2>/dev/null ; do sleep 0.1 ; done ; \\\n");
+    fprintf(f, "\t  /bin/ln -f $(MODULE).new $@ ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).lock\n");
+    fprintf(f, "\n");
+    // build .o from .cpp
+    fprintf(f, "$(MODULE).o: $(MODULE).cpp\n");
+    fprintf(f, "\t$(CXX) $(CXXFLAGS) -o $@ -c $< 2> $(MODULE).err && \\\n");
+    fprintf(f, "\t  $(CXX) $(CXXFLAGS) $(LIBFLAGS) -o $(MODULE).tmp $(MODULE).o 2>> $(MODULE).err ; \\\n");
+    fprintf(f, "\tstatus=$$? ; \\\n");
+    fprintf(f, "\tif [ $$status -ne 0 ] || [ -z $(MODULE).tmp ]; then \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).new ; \\\n");
+    fprintf(f, "\t  echo FAIL ; exit 1 ; \\\n");
+    fprintf(f, "\tfi\n");
+    fprintf(f, "\t-[ -f $(MODULE).lockX ] || /usr/bin/touch $(MODULE).lockX ; \\\n");
+    fprintf(f, "\t  while ! /bin/ln $(MODULE).lockX $(MODULE).lock 2>/dev/null ; do sleep 0.1 ; done ; \\\n");
+    fprintf(f, "\t  /bin/ln -f $(MODULE).tmp $(MODULE).new ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).tmp ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).err ; \\\n");
+    fprintf(f, "\t  /bin/rm -f $(MODULE).lock\n");
+    fprintf(f, "\n");
+    fclose(f);
+  }
+
+public:
+  bool try_to_start_build() {
+    // Try to create the .new file
+    std::lock_guard<NativeModule> mylock(*parent_);
+    int fd = ::open(hm_new_.c_str(), O_WRONLY|O_CREAT|O_EXCL, 0666);
+    if (fd < 0) {
+      // We lost the race to start the build
+      return false;
+    }
+    ::close(fd);
+    // The .new file may look the same age as the .cpp file, but
+    // the makefile is written to ignore the .new timestamp
+    write_mak();
+    write_cpp();
+    std::stringstream ss;
+    ss << "/usr/bin/make -B -C " << config.module_dir_ << " -f " << hm_mak_;
+    ss << " 1>/dev/null &";
+    int rc = system(ss.str().c_str());
+    if (rc < 0) perror("system");
```

Comment:
If we failed to execute the call shouldn't we crash? That means we could not compile some hail code, that doesn't seem recoverable?

Reply:
No, it's recoverable in the sense that "failed to compile code" will (eventually) give you a
NativeModule, but then when you query that with findOrBuild or any find*FuncN method,
you'll get back an error in the NativeStatus and can decide how to proceed.

I don't have it this way at the moment, but you could write the RowStore encoder in a way
which would try to build the C++ decoder, but then if the compile failed it would back off
to using the Java bytecode decoder.  *Probably* not what we want, but an example of how
there could sometimes be ways to recover and even complete the query.

I guess we could make that failure happen quicker by changing the build_state_
immediately.  But I don't really know how I would test that.

Reply:
Oh, god, please don't do that.  As soon as this goes live let's nuke the JVM decoder.  "I burn my bridges to light my way forward."

Reply:
Well yes.  I was expecting that when we got to code review of the RowStore.scala, there would
be a push to rip out all the bytecode stuff and shrink it down.  And I'll be cool with that.  If it works,
it should be the One True Way, and if it doesn't work, it shouldn't be merged.